File: 03월 25일 - 백트래킹/과제/17136.cpp
#include <iostream>
#include <vector>

using namespace std;
const int SIZE = 10;
const int INF = 26;

vector<vector<int>> paper(SIZE, vector<int>(SIZE));
vector<int> paper_cnt = {0, 5, 5, 5, 5, 5}; //인덱스 = 색종이 크기

bool promising(int row, int col, int paper_size) {
    for (int i = row; i < row + paper_size; i++) { //색종이 크기 만큼 돌면서 모두 1인지 확인
        for (int j = col; j < col + paper_size; j++) {
            if (paper[i][j] == 0) {
                return false;
            }
        }
    }
    return true;
}

void fill_paper(int row, int col, int paper_size, int fill_value) {
    for (int i = row; i < row + paper_size; i++) {
        for (int j = col; j < col + paper_size; j++) {
            paper[i][j] = fill_value;
        }
    }
}

int backtracking(int idx, int count) { //row: 행, col: 열, count: 색종이 개수
    int ans = INF;

    if (idx == SIZE * SIZE) { //색종이 다 붙임
        return min(ans, count);
    }
    int row = idx / SIZE;
    int col = idx % SIZE;

    if (!paper[row][col]) { //0이라면 -> 색종이 안 붙이므로 넘어감
        return backtracking(idx + 1, count);
    }
    if (ans <= count) { //기존 정답이 더 작으므로 count 값 답 가능성 x
        return INF;
    }

    for (int i = 5; i >= 1; i--) { //5 ~ 1까지 색종이 다 붙여보기
        if (row + i > SIZE || col + i > SIZE || !paper_cnt[i]) {
            continue;
        }
        if (promising(row, col, i)) {
            paper_cnt[i]--;
            fill_paper(row, col, i, 0);
            ans = min(ans, backtracking(idx + 1, count + 1));
            paper_cnt[i]++;
            fill_paper(row, col, i, 1);
        }
    }
    return ans;
}

/**
 * 5x5 색종이부터 1x1 색종이까지 모두 붙여보기
 * -> 필요한 색종이의 "최소" 개수이므로 가장 크기가 큰 5x5부터 시작
 *
 * [가지치기]
 * - 탐색을 하다 아직 색종이를 다 붙이지 않았는데, 개수가 현재 저장된 정답보다 크다면 탐색 종료
 */

int main() {
    //입력
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            cin >> paper[i][j];
        }
    }

    //연산
    int ans = backtracking(0, 0);

    //출력
    cout << (ans == INF ? -1 : ans); //간단한 출력에선 삼항연산자 사용해도 가독성 좋음
    return 0;
}

File: 03월 25일 - 백트래킹/과제/17136.py
import sys
input = sys.stdin.readline

"""
1. 붙일 수 있는 색종이의 최소 개수를 구하는 것이므로, 큰 색종이부터 붙여가면서 세어보아야 함
2. 색종이의 개수가 각 5장씩으로 제한되어 있기 때문에, 사용한 색종이의 개수를 기록해야 함
"""

SIZE = 10
MAX = 26

# (r, c)부터 시작해서 paper_size크기의 색종이를 붙일 수 있는지 확인하는 함수
def promising(r, c, paper_size):
    for i in range(r, r + paper_size):
        for j in range(c, c + paper_size):
            if board[i][j] == 0:
                return False
    return True

# board의 (r, c)부터 시작해서 paper_size크기를 num으로 채우는 함수
def fill_board(r, c, paper_size, num):
    for i in range(r, r + paper_size):
        for j in range(c, c + paper_size):
            board[i][j] = num
    return

def backtracking(idx, count):
    global answer   # 전역변수 사용
    
    if count > answer:
        return

    if idx == SIZE * SIZE:
        answer = min(answer, count)
        return

    x, y = idx // SIZE, idx % SIZE  # x행 y열
    
    if board[x][y] == 0:    # 현재 칸이 0이라면 넘어감
        backtracking(idx + 1, count)
        return

    for i in range(5, 0, -1):
        if x + i > SIZE or y + i > SIZE:
            continue
        if promising(x, y, i) and paper_cnt[i] > 0:
            paper_cnt[i] -= 1
            fill_board(x, y, i, 0)
            backtracking(idx + i, count + 1)
            paper_cnt[i] += 1
            fill_board(x, y, i, 1)

    return

# 입력
board = [list(map(int, input().split())) for _ in range(SIZE)]

paper_cnt = [5] * 6 # 남은 색종이의 수 (index 1~5 사용)
answer = MAX    # 최솟값 갱신할 변수

# 연산
backtracking(0, 0)

if answer == MAX: # 불가한 경우
    print(-1)
else:
    print(answer)